x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+    // See https://github.com/CosmWasm/cosmwasm/issues/1477
+    let ibc_transfer = MsgTransfer {
+        source_port: "transfer".to_string(),
+        source_channel: swap_msg_state.forward_to.channel.clone(),
+        token: Some(
+            Coin::new(
+                swap_response.amount.into(),
+                swap_response.token_out_denom.clone(),
+            )
+            .into(),
+        ),
+        sender: contract_addr.to_string(),
+        receiver: swap_msg_state.forward_to.receiver.clone().into(),
+        timeout_height: None,
+        timeout_timestamp: Some(ts.nanos()),
+        memo: memo_str,
+    };
+
+    // Base response
+    let response = Response::new()
+        .add_attribute("status", "ibc_message_created")
+        .add_attribute("ibc_message", format!("{:?}", ibc_transfer));
+
+    if !config.track_ibc_callbacks || swap_msg_state.forward_to.failed_delivery.is_none() {
+        // If we're not tracking callbacks, or there isn't any recovery addres,
+        // then there's no need to listen to the response of the send.
+        // The response data
+
+        // Add the response data since it won't be added in the reply.
+        let amount = swap_response.amount;
+        let denom = swap_response.token_out_denom;
+        let channel_id = swap_msg_state.forward_to.channel;
+        let to_address = swap_msg_state.forward_to.receiver;
+        let data = CrosschainSwapResponse {
+            msg: format!("Sent {amount}{denom} to {channel_id}/{to_address}"),
+        };
+
+        return Ok(response
+            .set_data(to_binary(&data)?)
+            .add_message(ibc_transfer));
+    }
+
+    // Check that there isn't anything stored in FORWARD_REPLY_STATES. If there
+    // is, it means that the contract is already waiting for a reply and should
+    // not override the stored state. This should never happen here, but adding
+    // the check for safety. If this happens there is likely a malicious attempt
+    // modify the contract's state before it has replied.
+    if FORWARD_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store the ibc send information and the user's failed delivery preference
+    // so that it can be handled by the response
+    FORWARD_REPLY_STATES.save(
+        deps.storage,
+        &ForwardMsgReplyState {
+            channel_id: swap_msg_state.forward_to.channel,
+            to_address: swap_msg_state.forward_to.receiver.into(),
+            amount: swap_response.amount.into(),
+            denom: swap_response.token_out_denom,
+            failed_delivery: swap_msg_state.forward_to.failed_delivery,
+        },
+    )?;
+
+    Ok(response.add_submessage(SubMsg::reply_on_success(ibc_transfer, FORWARD_REPLY_ID)))
+}
+
+use ::prost::Message; // Proveides ::decode() for MsgTransferResponse
```

Comment:
(nit) should this be at top of file?

Reply:
clippy doesn't care. I like importing trait implementations close to where they're used (because one of my pet peeves with rust is not knowing where some impls come from like in this one)